lib/spack/spack/directives.py

```diff
@@ -932,6 +982,244 @@ def _execute_requires(pkg: Type[spack.package_base.PackageBase]):
+    name = ""
+
+    def __init__(self, when):
+        self.removal_when = _make_when_spec(when)
+
+    @abstractmethod
+    def add_to_filtered(self, filtered, when, directive_entry):
+        pass
+
+    @abstractmethod
+    def get_directive(self, directive_entry):
+        pass
+
+    def remove(self):
+        removal_when = self.removal_when
+        if not removal_when:
+            return
+        complement_versions = removal_when.versions.complement()
+
+        def _remove(pkg):
+            removal_directive = self.make_directive(pkg)
+            directive_dict = getattr(pkg, self.name)
+            filtered = self.filter_directives(
+                complement_versions, directive_dict, removal_directive, removal_when
+            )
+            directive_dict.clear()
+            directive_dict.update(filtered)
+
+        return _remove
+
+    def filter_directives(
+        self, complement_versions, directive_dict, removal_directive, removal_when
+    ):
+        filtered = {}
+        for when, directives in directive_dict.items():
+            for directive_entry in self.iterate_directives(directives):
+                if self.get_directive(directive_entry) == removal_directive:
+                    if when == removal_when:
+                        continue
+                    elif when.versions.intersects(removal_when.versions):
+                        self.handle_intersection(
+                            when, directive_entry, complement_versions, filtered
+                        )
+                    else:
+                        self.add_to_filtered(filtered, when, directive_entry)
+                else:
+                    self.add_to_filtered(filtered, when, directive_entry)
+        return filtered
+
+    def handle_intersection(self, when, directive_entry, complement_versions, filtered):
+        new_when = when.copy()
+        new_when.versions = when.versions.intersection(complement_versions)
+        if new_when.versions != spack.version.VersionList():
+            self.add_to_filtered(filtered, new_when, directive_entry)
+
+    def iterate_directives(self, directives):
+        return directives
+
+
+class DropConflicts(DropDirectiveBase):
+    name = "conflicts"
+
+    def __init__(self, spec, when):
+        DropDirectiveBase.__init__(self, when)
+        self.make_directive = lambda pkg: spack.spec.Spec(spec)
+
+    def add_to_filtered(self, filtered, when, directive_entry):
+        filtered.setdefault(when, []).append(directive_entry)
+
+    def get_directive(self, directive_entry):
+        return directive_entry[0]
+
+
+class DropDependsOn(DropConflicts):
+    name = "dependencies"
+
+    def add_to_filtered(self, filtered, when, directive_entry):
+        name, directive = directive_entry
+        filtered.setdefault(when, {})[name] = directive
+
+    def iterate_directives(self, directives):
+        return directives.items()
+
+    def get_directive(self, directive_entry):
```

Comment:
Reading this, I'm curious about the following case:

```
class Foo(Package):
    depends_on("mpi@1:")

class Bar(Foo):
    drop_depends_on("mpi")
```
It appears that in this case `bar` would still depend on `mpi@1`, is that correct?

We should consider whether it's better to have spec satisfaction as the matching function for these cases. It seems more natural that way, although it does make it impossible to wipe out a general constraint without also wiping the more specific constraints.